sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuRoundRobinPartitioning.scala

```diff
@@ -42,21 +43,35 @@ case class GpuRoundRobinPartitioning(numPartitions: Int)
 
   def partitionInternal(batch: ColumnarBatch): (Array[Int], Array[GpuColumnVector]) = {
     val sparkTypes = GpuColumnVector.extractTypes(batch)
-    withResource(GpuColumnVector.from(batch)) { table =>
-      if (numPartitions == 1) {
+    if (1 == numPartitions) {
+      // Skip retry since partition number = 1
+      withResource(GpuColumnVector.from(batch)) { table =>
         val columns = (0 until table.getNumberOfColumns).zip(sparkTypes).map {
-          case(idx, sparkType) =>
-            GpuColumnVector.from(table.getColumn(idx).incRefCount(), sparkType)
+          case (idx, sparkType) =>
+            GpuColumnVector
+              .from(table.getColumn(idx).incRefCount(), sparkType)
         }.toArray
-        return (Array(0), columns)
+        (Array(0), columns)
       }
-      withResource(table.roundRobinPartition(numPartitions, getStartPartition)) { partedTable =>
-        val parts = partedTable.getPartitions
-        val columns = (0 until partedTable.getNumberOfColumns.toInt).zip(sparkTypes).map {
-          case(idx, sparkType) =>
-            GpuColumnVector.from(partedTable.getColumn(idx).incRefCount(), sparkType)
-        }.toArray
-        (parts, columns)
+    } else {
+      // Increase ref count since the caller will close the batch.
+      val spillableBatch = SpillableColumnarBatch(
+        GpuColumnVector.incRefCounts(batch), SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
```

Comment:
We should never be incrementing refcounts for batches sent to `withRetry` or `withRetryNoSplit`.  The point of these methods is to have _everything_ be spillable at that point so the memory framework is allowed to fully release all memory for a task at this control point (by spilling all of its memory).

Incrementing refcounts means the memory is _not_ spillable and violates that premise.  If the caller of this function is going to close this batch, then it seems like the caller of this function is where we need to perform the retry, not here.

Reply:
Updated. Since it's a helper method, close the batch in the callee side instead.